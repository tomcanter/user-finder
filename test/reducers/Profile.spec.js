import profileReducer, {
  getFromCache,
} from 'reducers/Profile';

describe('Reducer: profile', () => {

  describe('when passed no state', () => {
    it('should return the initial state', () => {
      expect(
        profileReducer(undefined, {})
      ).toMatchSnapshot();
    });
  });

  describe('when a profile request is pending', () => {
    it('should set isPending to true', () => {
      const beforeState = {
        isPending: false,
      };
      const afterState = profileReducer(beforeState, {type: 'PROFILE_REQUEST'});

      expect(
        afterState
      ).toMatchSnapshot();
    });
  });

  describe('when a profile request is successful', () => {
    describe('and the request is not cached', () => {
      it('should update the state with the userProfile and cache it', () => {
        const beforeState = {
          isPending: true,
          userProfile: {},
          cache: {},
        };
        const action = {
          profile: {
            login: 'tomcanter',
            name: 'Tom Canter',
          },
          type: 'PROFILE_SUCCESS',
        };
        const afterState = profileReducer(beforeState, action);

        expect(
          afterState
        ).toMatchSnapshot();
        expect(afterState).not.toBe(beforeState);
      });
    });

    describe('and the query is already cached', () => {
      it('should not merge the cached query back into the cache again', () => {
        const beforeState = {
          error: null,
          isPending: true,
          userProfile: {
            login: 'tomcanter',
          },
          cache: {
            tomcanter: {login: 'tomcanter'},
          },
        };
        const action = {
          profile: {login: 'tomcanter', test: 'foo'},
          type: 'PROFILE_SUCCESS',
        };
        const afterState = profileReducer(beforeState, action);

        expect(
          afterState
        ).toMatchSnapshot();
        expect(afterState).not.toBe(beforeState);
      });
    });
  });

  describe('when a search request fails', () => {
    it('should update the state with the error message', () => {
      const beforeState = {
        error: null,
        isPending: true,
      };
      const action = {
        response: {},
        message: 'The server died',
        type: 'PROFILE_FAILURE',
      };
      const afterState = profileReducer(beforeState, action);

      expect(
        afterState
      ).toMatchSnapshot();
      expect(afterState).not.toBe(beforeState);
    });
  });

  describe('getFromCache selector', () => {
    it('should return the cached value', () => {
      const store = {
        profile: {
          cache: {
            tom: {foo: 'bar'},
          },
        },
      };
      expect(getFromCache('tom')(store)).toEqual({foo: 'bar'});
    });
  });

});
