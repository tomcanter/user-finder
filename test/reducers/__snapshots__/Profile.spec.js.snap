// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer: profile when a profile request is pending should set isPending to true 1`] = `
Object {
  "isPending": true,
}
`;

exports[`Reducer: profile when a profile request is successful and the query is already cached should not merge the cached query back into the cache again 1`] = `
Object {
  "cache": Object {
    "tomcanter": Object {
      "login": "tomcanter",
    },
  },
  "error": null,
  "isPending": false,
  "userProfile": Object {
    "login": "tomcanter",
    "test": "foo",
  },
}
`;

exports[`Reducer: profile when a profile request is successful and the request is not cached should update the state with the userProfile and cache it 1`] = `
Object {
  "cache": Object {
    "tomcanter": Object {
      "login": "tomcanter",
      "name": "Tom Canter",
    },
  },
  "error": null,
  "isPending": false,
  "userProfile": Object {
    "login": "tomcanter",
    "name": "Tom Canter",
  },
}
`;

exports[`Reducer: profile when a search request fails should update the state with the error message 1`] = `
Object {
  "error": Object {
    "message": "The server died",
    "response": Object {},
  },
  "isPending": false,
}
`;

exports[`Reducer: profile when passed no state should return the initial state 1`] = `
Object {
  "cache": Object {},
  "error": null,
  "isPending": false,
  "userProfile": Object {},
}
`;
